File: chroma-league/src/main/java/net/booone/chromaleague/hud/animations/LoadingAnimation.java
package net.booone.chromaleague.hud.animations;

import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.Frame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.Color;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.sdk.ProgressiveRzKeySelector;
import net.booone.razersdk.sdk.RzKey;
import net.booone.razersdk.sdk.RzKeySelector;
import net.booone.razersdk.sdk.json.request.KeyboardEffect;

import java.util.stream.IntStream;

public class LoadingAnimation extends AnimatedFrame {
    private static final Color LOADING_COLOR = StaticColor.GREEN;
    private static final ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector(IntStream.concat(
                    IntStream.range(0, KeyboardEffect.KEYBOARD_COLUMNS),
                    IntStream.range(0, KeyboardEffect.KEYBOARD_COLUMNS)
                            .map(i -> KeyboardEffect.KEYBOARD_COLUMNS - i - 1))
            .mapToObj(col -> new RzKeySelector().withColumn(col).withRowBetween(RzKey.RZKEY_ESC, RzKey.RZKEY_LCTRL).asSet())
            .toList(), 2);

    @Override
    public Frame getFrame() {
        addAnimationFrame(new SimpleFrame(progressiveRzKeySelector.getNextPart(), LOADING_COLOR));
        return super.getFrame();
    }
}

File: src/hud/animations/LoadingAnimation.ts
import {AnimatedFrame} from '../../razer-sdk/animation/AnimatedFrame.js';
import {Frame} from '../../razer-sdk/animation/Frame.js';
import {SimpleFrame} from '../../razer-sdk/animation/SimpleFrame.js';
import {Color} from '../../razer-sdk/color/Color.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';
import {ProgressiveRzKeySelector} from '../../razer-sdk/sdk/ProgressiveRzKeySelector.js';
import {RzKey} from '../../razer-sdk/sdk/RzKey.js';
import {KEYBOARD_COLUMNS, RzKeySelector} from '../../razer-sdk/sdk/RzKeySelector.js';

const LOADING_COLOR: Color = StaticColor.GREEN;

function buildParts(): ReadonlySet<RzKey>[] {
    const forward: ReadonlySet<RzKey>[] = [];
    for (let col = 0; col < KEYBOARD_COLUMNS; col++) {
        forward.push(
            new RzKeySelector()
                .withColumn(col)
                .withRowBetween(RzKey.RZKEY_ESC, RzKey.RZKEY_LCTRL)
                .asSet(),
        );
    }
    const backward: ReadonlySet<RzKey>[] = [];
    for (let col = KEYBOARD_COLUMNS - 1; col >= 0; col--) {
        backward.push(
            new RzKeySelector()
                .withColumn(col)
                .withRowBetween(RzKey.RZKEY_ESC, RzKey.RZKEY_LCTRL)
                .asSet(),
        );
    }
    return [...forward, ...backward];
}

const progressiveRzKeySelector = new ProgressiveRzKeySelector(buildParts(), 2);

export class LoadingAnimation extends AnimatedFrame {
    pollFrame(): Frame {
        this.addAnimationFrame(new SimpleFrame([...progressiveRzKeySelector.getNextPart()], LOADING_COLOR));
        return super.pollFrame();
    }
}
